[REEF-1073] Improve ConstructorDefImpl.equalsIgnoreOrder algorithm

dongjoon-hyun opened the pull request:
JIRA:
  [REEF-1073](https://issues.apache.org/jira/browse/REEF-1073)

Pull Request:
  This closes #

dongjoon-hyun commented:
This PR includes `testMicroBenchmark` just for review. After review, I'll replace `equals` with `equalsNew` and delete `testMicroBenchmark`.

tcNickolas commented on lang/java/reef-tang/tang/src/main/java/org/apache/reef/tang/implementation/types/ConstructorDefImpl.java:
```diff
@@ -129,6 +130,48 @@ public boolean equals(final Object o) {
+   * @return
+   */
+  private boolean equalsIgnoreOrderUsingHashMap(final ConstructorDef<?> def) {
+    HashMap map = new HashMap();
```
What's the benefit of using HashMap instead of HashSet? There should be not much performance difference, but readability-wise set looks better, as we don't need values, only keys

tcNickolas commented on lang/java/reef-tang/tang/src/main/java/org/apache/reef/tang/implementation/types/ConstructorDefImpl.java:
```diff
@@ -129,6 +130,48 @@ public boolean equals(final Object o) {
+      if (!map.containsKey(getArgs()[i].getName())) {
+        return false;
+      }
+    }
```
The above code seems better when written as

~~~
    for (ConstructorArg a : def.getArgs()) {
      map.put(a.getName(), null);
    }
    for (ConstructorArg a : getArgs()) {
      if (!map.containsKey(a.getName())) {
        return false;
      }
    }
~~~

It requires less calls to getArgs() and (arguably) seems a little faster when I run benchmark locally.

tcNickolas commented on lang/java/reef-tang/tang/src/main/java/org/apache/reef/tang/implementation/types/ConstructorDefImpl.java:
```diff
@@ -129,6 +130,48 @@ public boolean equals(final Object o) {
+    ConstructorDef<?> def = (ConstructorDef<?>) o;
+    if (length != def.getArgs().length) {
+      return false;
+    } else if (length == 0) {
```
I think we don't need `else`'s after `if` which ends in `return`. Separate `if`'s would be more readable

tcNickolas commented:
I've done a pass. The logic behind the change looks good.

dongjoon-hyun commented:
Thank you, @tcNickolas . I've changed according to your comments except `HashMap`. 

I chose `HashMap` because it's more primitive class. As you know, `HashSet` uses `HashMap`, so `HashSet` has more overhead than `HashMap` in terms of CPU/Memory. After updating the code according to your comments, I made and tested `HashSet` version multiple times again. The result shows `HashSet` version does not outperform the original `for-loop` version sometimes for some `n`.

Although there is a readability issue, I think this patch is meaningful only when it outperforms the original version.

And, I rebased and clean up the code for merging. For the previous patch diff can be found in the following link.

https://app.box.com/s/4fir03lpw5y4jgh847c8tm0jan851nss

tcNickolas commented:
I agree, performance is more important than readability in this case, I didn't realize there would be so much overhead.

Looks good and passes tests, I'll merge

dongjoon-hyun commented:
Thank you, @tcNickolas !
